### ISSUE TITLE
Allow for album covers to be retrieved from webserver in lan

### ISSUE THREAD

**Issue by rhoeden** [gateship-one/malp#13]
I'm using an nginx instance in conjunction with mpd to serve high quality (1500x1500) album covers in my lan. Previously I used  [mpdroid](https://github.com/abarisain/dmix) that has this function built-in and I was hoping this could be something worth implementing.

```
$ cat mpd-covers.conf

server {
    listen 80;
    server_name 10.0.0.110;

    location / {
        root /media/music;
        rewrite /(.*) /$1 break;
        try_files $uri $uri /.no-cover-art.jpg;
        allow   10.0.0.0/24;
        deny    all;
    }
}
```

**Comment by djselbeck** [comment]
It should be possible to implement. But I've to check how to configure it and what option to present to the user.

**Comment by hansheijmans** [comment]
Any news on this?

**Comment by djselbeck** [comment]
Not yet, at the moment I'm working on streaming and this is not so important because the MusicBrainz database works good for me at the moment. I also have no way to test this for now.

**Comment by transiti0nary** [comment]
Is this feature still being considered for implementation?

**Comment by djselbeck** [comment]
Yes, but for personal reasons we don't have a lot of time for Android development at the time

**Comment by transiti0nary** [comment]
No problem, I may have a go at implementing it myself if I get some spare time (which probably isn't that likely haha) ... do you think the LAN artwork retrieval code from the (abandoned, AFAIK) MPDroid could be easily adapted to work with MALP or better to start from scratch?

**Comment by djselbeck** [comment]
I started to implement it in https://github.com/gateship-one/malp/tree/http-cover this branch.

MPDroid is not usable for me as I want all code to be under the GPL license.

**Comment by hansheijmans** [comment]
As I can see in my nginx logs, mpdroid is looking for all kinds of picture files like cover.jpg, front.jpg, folder.jpg, <name_of_album>.jpg, also *.png etc. Also I like the possiblity of using MPD-hosts based on WLAN like mpdroid does.
Would you be able to do the same?

Just a thought...

**Comment by transiti0nary** [comment]
More than sufficient, that's perfect (for me at least)!

**Comment by transiti0nary** [comment]
Btw, when approx. will this feature be pushed to Play / F-Droid or .apk from repo?

**Comment by djselbeck** [comment]
This year probably 👅 , after some testing

**Comment by gitteric** [comment]
Thanks for taking care of this feature, was really looking forward to it!
I did some tests and found two glitches of which one is not directly related to the http cover fetching:

- Bulk cover downloading failed on larger libraries. Mine is appx. 50.000 tracks and the mpd command "listallinfo" takes longer than five seconds, which let's "waitForResponse" throw an exception
- My server failed to serve some covers due to screwed up special chars in URI. It seemed that "encodeURLUnsafeCharacters" was not able handle chars resulting in representations > 1 byte

I fixed (or at least tried to fix) both in my fork, see commit:
[https://github.com/gitteric/malp/commit/790a52421a86333277639e3c0833ec1280377911](url)
Works for me ;)

Happy new year!

**Comment by transiti0nary** [comment]
Would this implementation be able to work with volumio as is outlined here? https://volumio.github.io/docs/User_Manual/Configure_MPD_Clients_for_albumart.html (rewrites the actual path as /cover-art/)

**Comment by djselbeck** [comment]
I wouldn't know. I don't support volumio.

**Comment by transiti0nary** [comment]
I think it is doing the same thing as the guide on MPDroid wiki section to rewrite your music folder to /cover-art/ for nginx server. I think it would just require being able to specify the root folder to search for cover files rather than to default to the actual music library path?

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by djselbeck** [comment]
MALP can look in every path containing the path of a song (%d is replaced with the path) everything before the %d is possible. e.g: http://blablaHost/moreBla/%d

**Comment by transiti0nary** [comment]
That's perfect. Thanks so much for taking the time to implement this! I noticed this has been added to the readme as a feature, is it available in the latest release 1.1.12? If so how is it enabled, simply turn artwork provider to off?

**Comment by djselbeck** [comment]
MALP can only retrieve covers from the same path as the music is located. For future versions I can add more %placeholders for artist & album

**Comment by transiti0nary** [comment]
The covers are in the same directory as the albums themselves, and I have made the parent directory of my library available on my nginx server.

For example the album Amnesiac by Radiohead is in /home/pi/hdd/music/Radiohead/Amnesiac%20(2001)/ and if I navigate to http://my.mpd.server:80/home/pi/hdd/music/Radiohead/Amnesiac%20(2001)/cover.jpg I see the album art for Amnesiac. All the .flac files are in that folder also.

Will this setup not currently work?

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by djselbeck** [comment]
If you mpd database directory is set to "/home/pi/hdd/music" then the path for MALPs artwork should be: "http://my.mpd.server/home/pi/hdd/music/%d"

**Comment by transiti0nary** [comment]
Ah I understand now, thanks! Seems to be working 👍

**Comment by transiti0nary** [comment]
I have noticed one 'bug' (though maybe it is by design for some reason); MALP does not seem to fetch the covers until a track from the album is actually played?

It would be awesome if it automatically fetched them when the album is displayed on screen, or if there was a 'download all' button in the settings maybe.

Otherwise this feature works perfectly and I haven't had any other problems, thanks so much for adding it!

-------- Original Message --------

**Comment by rodgersan** [comment]
Hello,
As mentioned by @gitteric URLs containing special characters do not work. All covers were fetched except for 3 paths containing "ä" and "é".
Thanks for this wonderful app!
